File: sdk/typescript/src/transport.ts
import { ConnectionError, RpcError, SessionReplacedError } from "./errors.ts";

export interface RpcTransport {
  call(method: string, params: unknown): Promise<unknown>;
  /** Send a JSON-RPC call followed by binary frame(s), wait for JSON response. */
  callWithBinary(method: string, params: unknown, data: Uint8Array): Promise<unknown>;
  /** Send a JSON-RPC call and collect binary frames before the JSON response. */
  callExpectBinary(method: string, params: unknown): Promise<{ result: unknown; binary: Uint8Array[] }>;
  notify(method: string, params: unknown): void;
  onNotification(handler: (method: string, params: unknown) => void): void;
  onReplaced(handler: () => void): void;
  close(): Promise<void>;
  sendBinary(data: Uint8Array): void;
  onBinary(handler: (data: Uint8Array) => void): void;
}

interface PendingRequest {
  resolve: (value: unknown) => void;
  reject: (reason: unknown) => void;
  /** If set, binary frames are collected here before the JSON response arrives. */
  binaryBufs?: Uint8Array[];
  expectBinary?: boolean;
}

const MAX_FRAME_SIZE = 10 * 1024 * 1024; // 10MB
const CHUNK_SIZE = 1 * 1024 * 1024; // 1MB

export class WsTransport implements RpcTransport {
  private ws: WebSocket | null = null;
  private nextId = 1;
  private pending = new Map<number, PendingRequest>();
  private notificationHandlers: Array<(method: string, params: unknown) => void> = [];
  private replacedHandlers: Array<() => void> = [];
  private binaryHandler: ((data: Uint8Array) => void) | null = null;
  private connectPromise: Promise<void> | null = null;
  /** ID of the request currently expecting binary frames, or 0. */
  private binaryExpectId = 0;

  async connect(url: string, headers?: Record<string, string>): Promise<void> {
    if (this.connectPromise) return this.connectPromise;

    this.connectPromise = new Promise<void>((resolve, reject) => {
      try {
        this.ws = new WebSocket(url, headers ? Object.entries(headers).map(([k, v]) => `${k}:${v}`) : undefined);
      } catch {
        // Bun's WebSocket accepts headers directly in some configurations.
        // Fall back to no subprotocol if the above fails.
        this.ws = new WebSocket(url);
      }

      this.ws.binaryType = "arraybuffer";

      this.ws.onopen = () => resolve();

      this.ws.onerror = (ev) => {
        const msg = ev instanceof ErrorEvent ? ev.message : "WebSocket error";
        reject(new ConnectionError(msg));
      };

      this.ws.onclose = (ev) => {
        // Reject all pending requests.
        for (const [, p] of this.pending) {
          p.reject(new ConnectionError(`WebSocket closed: ${ev.code} ${ev.reason}`));
        }
        this.pending.clear();
      };

      this.ws.onmessage = (ev) => {
        if (ev.data instanceof ArrayBuffer) {
          const data = new Uint8Array(ev.data);

          // Route to the pending request expecting binary frames.
          if (this.binaryExpectId > 0) {
            const p = this.pending.get(this.binaryExpectId);
            if (p && p.expectBinary) {
              if (!p.binaryBufs) p.binaryBufs = [];
              p.binaryBufs.push(data);
              return;
            }
          }

          // Fall back to the one-shot binary handler.
          if (this.binaryHandler) {
            const handler = this.binaryHandler;
            this.binaryHandler = null;
            handler(data);
          }
          return;
        }

        let msg: { id?: number; method?: string; params?: unknown; result?: unknown; error?: { code: number; message: string } };
        try {
          msg = JSON.parse(ev.data as string);
        } catch {
          return;
        }

        // JSON-RPC response (has id).
        if (msg.id !== undefined) {
          const p = this.pending.get(msg.id);
          if (!p) return;
          this.pending.delete(msg.id);

          // Clear binary expect tracking if this was the binary-expecting request.
          if (this.binaryExpectId === msg.id) {
            this.binaryExpectId = 0;
          }

          if (msg.error) {
            p.reject(new RpcError(msg.error.message, msg.error.code));
          } else {
            p.resolve(msg.result);
          }
          return;
        }

        // JSON-RPC notification (no id).
        if (msg.method) {
          if (msg.method === "session.replaced") {
            for (const h of this.replacedHandlers) h();
            this.ws?.close();
            return;
          }

          for (const h of this.notificationHandlers) {
            h(msg.method, msg.params);
          }
        }
      };
    });

    return this.connectPromise;
  }

  call(method: string, params: unknown): Promise<unknown> {
    if (!this.ws || this.ws.readyState !== WebSocket.OPEN) {
      return Promise.reject(new ConnectionError("WebSocket not connected"));
    }

    const id = this.nextId++;
    return new Promise((resolve, reject) => {
      this.pending.set(id, { resolve, reject });
      this.ws!.send(JSON.stringify({ jsonrpc: "2.0", id, method, params }));
    });
  }

  callWithBinary(method: string, params: unknown, data: Uint8Array): Promise<unknown> {
    if (!this.ws || this.ws.readyState !== WebSocket.OPEN) {
      return Promise.reject(new ConnectionError("WebSocket not connected"));
    }

    const id = this.nextId++;
    const promise = new Promise((resolve, reject) => {
      this.pending.set(id, { resolve, reject });
    });

    // Send JSON-RPC request.
    this.ws.send(JSON.stringify({ jsonrpc: "2.0", id, method, params }));

    // Send binary data in chunks.
    for (let offset = 0; offset < data.byteLength; offset += CHUNK_SIZE) {
      const end = Math.min(offset + CHUNK_SIZE, data.byteLength);
      this.ws.send(data.subarray(offset, end));
    }

    return promise;
  }

  callExpectBinary(method: string, params: unknown): Promise<{ result: unknown; binary: Uint8Array[] }> {
    if (!this.ws || this.ws.readyState !== WebSocket.OPEN) {
      return Promise.reject(new ConnectionError("WebSocket not connected"));
    }

    const id = this.nextId++;
    this.binaryExpectId = id;

    return new Promise((resolve, reject) => {
      const pending: PendingRequest = {
        resolve: (result) => {
          resolve({ result, binary: pending.binaryBufs ?? [] });
        },
        reject,
        expectBinary: true,
        binaryBufs: [],
      };
      this.pending.set(id, pending);
      this.ws!.send(JSON.stringify({ jsonrpc: "2.0", id, method, params }));
    });
  }

  notify(method: string, params: unknown): void {
    if (!this.ws || this.ws.readyState !== WebSocket.OPEN) {
      throw new ConnectionError("WebSocket not connected");
    }
    this.ws.send(JSON.stringify({ jsonrpc: "2.0", method, params }));
  }

  onNotification(handler: (method: string, params: unknown) => void): void {
    this.notificationHandlers.push(handler);
  }

  onReplaced(handler: () => void): void {
    this.replacedHandlers.push(handler);
  }

  sendBinary(data: Uint8Array): void {
    if (!this.ws || this.ws.readyState !== WebSocket.OPEN) {
      throw new ConnectionError("WebSocket not connected");
    }
    // Send in chunks to stay within frame size limits.
    for (let offset = 0; offset < data.byteLength; offset += CHUNK_SIZE) {
      const end = Math.min(offset + CHUNK_SIZE, data.byteLength);
      const chunk = data.subarray(offset, end);
      if (chunk.byteLength > MAX_FRAME_SIZE) {
        throw new RpcError(`frame size ${chunk.byteLength} exceeds maximum ${MAX_FRAME_SIZE}`, -32000);
      }
      this.ws.send(chunk);
    }
  }

  onBinary(handler: (data: Uint8Array) => void): void {
    this.binaryHandler = handler;
  }

  async close(): Promise<void> {
    if (this.ws) {
      this.ws.close();
      this.ws = null;
    }
    this.pending.clear();
    this.connectPromise = null;
  }
}

File: sdk/python/byggflow/sandbox/transport.py
"""Transport abstraction for RPC communication with the daemon.

Provides both the abstract ``RpcTransport`` protocol and a concrete
``WsTransport`` implementation backed by the ``websockets`` library.
"""

from __future__ import annotations

import asyncio
import json
from abc import ABC, abstractmethod
from typing import Any, Callable, Dict, List, Optional, Tuple

from .errors import ConnectionError, RpcError, SessionReplacedError

MAX_FRAME_SIZE = 10 * 1024 * 1024  # 10MB
CHUNK_SIZE = 1 * 1024 * 1024  # 1MB


class RpcTransport(ABC):
    """Abstract base for daemon communication."""

    @abstractmethod
    async def send(self, op: str, params: Dict[str, Any]) -> Any:
        """Send an RPC request and return the result payload."""
        ...

    @abstractmethod
    async def call_with_binary(
        self, method: str, params: Dict[str, Any], data: bytes
    ) -> Any:
        """Send a JSON-RPC request followed by binary data chunks, return the JSON response."""
        ...

    @abstractmethod
    async def call_expect_binary(
        self, method: str, params: Dict[str, Any]
    ) -> Tuple[Any, List[bytes]]:
        """Send a JSON-RPC request and collect binary frames before the JSON response."""
        ...

    @abstractmethod
    async def send_binary(self, data: bytes) -> None:
        """Send raw binary data, chunking at CHUNK_SIZE."""
        ...

    @abstractmethod
    async def close(self) -> None:
        """Gracefully shut down the transport."""
        ...


class WsTransport(RpcTransport):
    """WebSocket-based JSON-RPC transport using the ``websockets`` library."""

    def __init__(self) -> None:
        self._ws: Any = None
        self._next_id: int = 1
        self._pending: Dict[int, asyncio.Future[Any]] = {}
        self._notification_handlers: list[Callable[[str, Any], None]] = []
        self._replaced_handlers: list[Callable[[], None]] = []
        self._binary_handler: Optional[Callable[[bytes], None]] = None
        self._read_task: Optional[asyncio.Task[None]] = None
        self._binary_expect_id: int = 0
        self._binary_bufs: Dict[int, list[bytes]] = {}

    async def connect(self, url: str, headers: Optional[Dict[str, str]] = None) -> None:
        """Open a WebSocket connection and start the read loop."""
        import websockets

        extra_headers = headers or {}
        self._ws = await websockets.connect(url, additional_headers=extra_headers)
        self._read_task = asyncio.create_task(self._read_loop())

    async def _read_loop(self) -> None:
        """Background task that reads messages from the WebSocket."""
        try:
            async for message in self._ws:
                if isinstance(message, bytes):
                    # Route to binary-expecting request if one is active.
                    if self._binary_expect_id != 0:
                        req_id = self._binary_expect_id
                        if req_id not in self._binary_bufs:
                            self._binary_bufs[req_id] = []
                        self._binary_bufs[req_id].append(message)
                    elif self._binary_handler is not None:
                        handler = self._binary_handler
                        self._binary_handler = None
                        handler(message)
                    continue

                # Text message — parse JSON-RPC.
                try:
                    msg = json.loads(message)
                except (json.JSONDecodeError, TypeError):
                    continue

                # Response (has id).
                if "id" in msg and msg["id"] is not None:
                    req_id = msg["id"]
                    future = self._pending.pop(req_id, None)
                    if future is None or future.done():
                        continue

                    if "error" in msg and msg["error"]:
                        err = msg["error"]
                        future.set_exception(
                            RpcError(err.get("message", "Unknown error"), err.get("code", -1))
                        )
                    else:
                        future.set_result(msg.get("result"))
                    continue

                # Notification (no id).
                method = msg.get("method")
                if method:
                    if method == "session.replaced":
                        for h in self._replaced_handlers:
                            h()
                        await self._ws.close()
                        return

                    params = msg.get("params")
                    for h in self._notification_handlers:
                        h(method, params)

        except Exception:
            # Connection closed or error — reject all pending requests.
            pass
        finally:
            for future in self._pending.values():
                if not future.done():
                    future.set_exception(ConnectionError("WebSocket closed"))
            self._pending.clear()

    async def send(self, op: str, params: Dict[str, Any]) -> Any:
        """Send a JSON-RPC request and wait for the response."""
        return await self.call(op, params)

    async def call(self, method: str, params: Any = None) -> Any:
        """Send a JSON-RPC request and return the result."""
        if self._ws is None:
            raise ConnectionError("WebSocket not connected")

        req_id = self._next_id
        self._next_id += 1

        loop = asyncio.get_running_loop()
        future: asyncio.Future[Any] = loop.create_future()
        self._pending[req_id] = future

        message = json.dumps({"jsonrpc": "2.0", "id": req_id, "method": method, "params": params})
        await self._ws.send(message)

        return await future

    async def notify(self, method: str, params: Any = None) -> None:
        """Send a JSON-RPC notification (no response expected)."""
        if self._ws is None:
            raise ConnectionError("WebSocket not connected")

        message = json.dumps({"jsonrpc": "2.0", "method": method, "params": params})
        await self._ws.send(message)

    def on_notification(self, handler: Callable[[str, Any], None]) -> None:
        """Register a callback for server-initiated notifications."""
        self._notification_handlers.append(handler)

    def on_replaced(self, handler: Callable[[], None]) -> None:
        """Register a callback for ``session.replaced`` notifications."""
        self._replaced_handlers.append(handler)

    async def send_binary(self, data: bytes) -> None:
        """Send binary data, chunking at CHUNK_SIZE."""
        if self._ws is None:
            raise ConnectionError("WebSocket not connected")
        offset = 0
        while offset < len(data):
            end = min(offset + CHUNK_SIZE, len(data))
            chunk = data[offset:end]
            await self._ws.send(chunk)
            offset = end

    async def call_with_binary(
        self, method: str, params: Dict[str, Any], data: bytes
    ) -> Any:
        """Send a JSON-RPC request, then send binary data in chunks, wait for JSON response."""
        if self._ws is None:
            raise ConnectionError("WebSocket not connected")

        req_id = self._next_id
        self._next_id += 1

        loop = asyncio.get_running_loop()
        future: asyncio.Future[Any] = loop.create_future()
        self._pending[req_id] = future

        message = json.dumps({"jsonrpc": "2.0", "id": req_id, "method": method, "params": params})
        await self._ws.send(message)

        # Send binary data in chunks.
        await self.send_binary(data)

        return await future

    async def call_expect_binary(
        self, method: str, params: Dict[str, Any]
    ) -> Tuple[Any, List[bytes]]:
        """Send a JSON-RPC request and collect binary frames arriving before the JSON response."""
        if self._ws is None:
            raise ConnectionError("WebSocket not connected")

        req_id = self._next_id
        self._next_id += 1

        loop = asyncio.get_running_loop()
        future: asyncio.Future[Any] = loop.create_future()
        self._pending[req_id] = future

        # Tell the read loop to collect binary frames for this request.
        self._binary_expect_id = req_id
        self._binary_bufs[req_id] = []

        message = json.dumps({"jsonrpc": "2.0", "id": req_id, "method": method, "params": params})
        await self._ws.send(message)

        result = await future

        # Collect and clean up binary buffers.
        self._binary_expect_id = 0
        bufs = self._binary_bufs.pop(req_id, [])

        return result, bufs

    def on_binary(self, handler: Callable[[bytes], None]) -> None:
        """Register a one-shot binary message handler."""
        self._binary_handler = handler

    async def close(self) -> None:
        """Close the WebSocket and cancel the read task."""
        if self._read_task is not None:
            self._read_task.cancel()
            try:
                await self._read_task
            except asyncio.CancelledError:
                pass
            self._read_task = None

        if self._ws is not None:
            await self._ws.close()
            self._ws = None

        self._pending.clear()
